### DISCUSSION TITLE
Lower minimum brightness

### DISCUSSION THREAD

**Discussion by JosefRypacek** [arendst/Tasmota#18639] [Ideas]
I'm using Shelly RGBW2 with multiple white LED strips and would like to reduce brightness for night even lower than 1%. But this would be similar for any other configuration of Tasmota light.

I'm using `LedTable 1` (it's even worse without it).

This could be solved using `option15` and using pure `pwm` commands (range 0-1023), but then you lost fading, HA compatibility and more...

I've found, that 1% is converted to 0-255 value => probably **3** and this is converted to 0-1023 value => probably **12**.
Then this number is modified by LedTable (gamma_table) and then is ac_dimmer_table.
The whole process is too complex and I don't know exact numbers.

I've created a simple patch to force 1% to be converted into 1/1023 pwm which results in lowest possible brightness.

I don't have idea how such a functionality could be implemented into Tasmota, but I hope my patch could be useful for anyone with needs for night light and brightness reduction.

https://github.com/JosefRypacek/Tasmota/commit/e5ef9a1c6bb2eb974b2e4be4a5cdf7c205882165?diff=unified

**Comment by s-hadinger** [comment]
Here are the tables used to convert from a Tasmota value to the actual PWM value (with 10 bits resolution):

From 8 bits:
```

//   0,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
//   1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  2,  2,  2,  2,  2,  2,
//   2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  3,  3,  3,  3,  3,  3,
//   3,  3,  3,  3,  3,  3,  4,  4,  4,  4,  5,  5,  5,  6,  6,  6,
//   6,  7,  7,  7,  7,  8,  8,  8,  9,  9,  9,  9, 10, 10, 10, 11,
//  11, 12, 12, 12, 13, 13, 14, 14, 15, 15, 16, 16, 17, 17, 17, 18,
//  18, 19, 19, 20, 20, 21, 21, 21, 22, 22, 23, 23, 24, 24, 25, 25,
//  25, 26, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 36, 37, 38,
//  39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 50, 51, 52, 53,
//  54, 55, 56, 57, 58, 59, 60, 61, 61, 62, 63, 64, 65, 67, 68, 69,
//  71, 72, 73, 75, 76, 78, 79, 80, 82, 83, 85, 86, 87, 89, 90, 91,
//  93, 94, 95, 97, 98,100,101,102,104,105,107,108,109,111,112,114,
// 116,118,120,122,124,125,127,129,131,133,135,137,139,141,143,144,
// 146,148,150,152,154,156,158,160,162,164,166,168,170,171,173,175,
// 178,180,183,185,188,190,193,195,198,200,203,205,208,210,213,215,
// 218,220,223,225,228,230,233,235,238,240,243,245,248,250,253,255
```

While when converting from Dimmer percentage 0..100, here are the converted values:
```
// Output for Dimmer 0..100 values
//   0,  1,  2,  3,  3,  4,  4,  5,  5,  6,  7,  7,  8,  8,  9,
//  10, 10, 11, 12, 12, 13, 15, 17, 21, 23, 26, 28, 31, 34, 37,
//  40, 43, 49, 52, 58, 61, 67, 70, 76, 79, 84, 90, 93, 99,102,
// 110,117,128,135,146,158,166,177,184,195,202,213,221,232,239,
// 250,261,272,289,300,317,328,344,356,372,389,400,417,428,444,
// 458,480,496,518,534,557,579,595,617,635,657,673,695,713,743,
// 773,793,823,843,873,893,923,943,973,993,1023
```

As you can see, the curve of gamma correction for Dimmer has been carefully crafted so that at low values it uses the whole 10 bits resolution.

Using `LedTable 1` and `Dimmer 1` actually gives you the lowest supported brightness of 1/1024. Do your measures differ?

**Comment by JosefRypacek** [comment]
You are partially right. With my low knowledge I don't fully understand, but I added logging into functions in `xdrv_04_light_utils.ino` and configured my device as follows.

```
SetOption68 1
Fade 0
```

Then got following output
```
21:15:41.453 CMD: channel1 1
21:15:41.493 change8to10: input=3, output=14
21:15:41.503 ledGamma10_10: input=14, output=1
21:15:41.541 ac_zero_cross_power: input=1, output=8
21:15:41.560 change10to8: input=1, output=1
```

I can confirm that real brightness is probably the same like using
```
setoption15 0
pwm1 8
```


I don't think this is something bad that should be fixed, but for my night light application i need real pwm 1/1023.
And the only required modification seems to be modification of `ac_dimmer_table`.

```
const gamma_table_t ac_dimmer_table[] = { 
  {     0,      0 },
  {     1,      1 }, // add this line to get lowest possible brightness
  {    10,     64 },
```

**Comment by s-hadinger** [comment]
I'm surprised. `ac_dimmer_table` is only used for zerocross AC dimmers, when setting `SetOption99 1`. I don't believe that the shelly RGBW2 is an AC dimmer, this wouldn't make sense to drive Leds.

**Reply by JosefRypacek** [reply]
```
08:03:35.325 CMD: SetOption99
08:03:35.331 MQT: stat/tasmota_287958/RESULT = {"SetOption99":"OFF"}
```

My Shelly is using 12V DC for both input and output for LED stripes. :)

I'm available to provide additional information or test any future code changes. :)

**Reply by s-hadinger** [reply]
Let's do it from a fresh start. Reset your device to factory settings and start over. Also share your configuration. I can't magically guess what you have already done.

`ac_dimmer_table` is only used in two cases: 1/ `SetOption99 1`, 2/ in xdrv_68 which is also enabled only with SetOption99 1

**Comment by JosefRypacek** [comment]
I did

- configuration reset
- setup Wi-FI
- set template - tried both [FIRST](https://templates.blakadder.com/shelly_RGBW2.html), [SECOND](https://tasmota.github.io/docs/devices/Shelly-RGBW2/#template)
- SetOption68 1

The result is still the same.

```
18:28:17.997 RSL: STATUS0 = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota","","",""],"Topic":"tasmota_287958","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0},"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:01:10","StartupUTC":"2023-05-15T17:27:07","Sleep":50,"CfgHolder":4617,"BootCount":10,"BCResetTime":"2023-05-15T18:12:29","SaveCount":43,"SaveAddress":"F5000"},"StatusFWR":{"Version":"12.5.0.2(tasmota)","BuildDateTime":"2023-05-14T20:01:10","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"358/699"},"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["xxxxx",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00040080","00006000","00004000","00000000"]},"StatusMEM":{"ProgramSize":634,"Free":368,"Heap":23,"ProgramFlashSize":1024,"FlashSize":2048,"FlashChipId":"154068","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000080"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62,68","Sensors":"1,2,3,4,5,6","I2CDriver":"7"},"StatusNET":{"Hostname":"tasmota-287958-6488","IPAddress":"xxxxx","Gateway":"xxxxx","Subnetmask":"255.255.255.0","DNSServer1":"xxxxx","DNSServer2":"0.0.0.0","Mac":"xxxxx","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0},"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_287958","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4},"StatusTIM":{"UTC":"2023-05-15T17:28:17","Local":"2023-05-15T18:28:17","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"05:08","Sunset":"20:24"},"StatusSNS":{"Time":"2023-05-15T18:28:17"},"StatusSTS":{"Time":"2023-05-15T18:28:17","Uptime":"0T00:01:10","UptimeSec":70,"Heap":22,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":101,"MqttCount":0,"POWER1":"ON","Channel1":1,"POWER2":"OFF","Channel2":1,"POWER3":"OFF","Channel3":0,"POWER4":"OFF","Channel4":0,"Color":"03030101","Wifi":{"AP":1,"SSId":"xxxxx","BSSId":"xxxxx","Channel":1,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:05"}}}
```

**Reply by s-hadinger** [reply]
Hmmm. It seems that the ac-dimmer lookup table is always used even for leds. This change was introduced here https://github.com/arendst/Tasmota/pull/17177

@stefanbode can you double check #17177
I have the impression that the test is reversed. It applies AC-dimmer when it is not enabled.

**Reply by stefanbode** [reply]
This is very strange because the call of the function is inside the dimmer. Should never be called on led. Normally. I'm quite busy right now and in the US. Maybe you can have a look

**Reply by s-hadinger** [reply]
Ok, let me propose a fix here.

**Comment by s-hadinger** [comment]
@JosefRypacek Thanks for reporting the problem. It appears to be a bug. It should be now fixed in the `development` branch. Now `Dimmer 1` will actually be the lowest possible brightness (1/1023).

See https://github.com/arendst/Tasmota/pull/18666

**Reply by JosefRypacek** [reply]
Thanks! I can confirm it's working correctly now.
